packages/mui-base/src/useNumberField/useNumberField.ts

```diff
@@ -0,0 +1,802 @@
+
+function handleGlobalContextMenu(event: Event) {
+  event.preventDefault();
+}
+
+const CHANGE_VALUE_TICK_DELAY = 60;
+const START_AUTO_CHANGE_DELAY = 400;
+const TOUCH_TIMEOUT = 50;
+const MAX_POINTER_MOVES_AFTER_TOUCH = 3;
+const SCROLLING_POINTER_MOVE_DISTANCE = 8;
+
+/**
+ *
+ * Demos:
+ *
+ * - [Number Field](https://mui.com/base-ui/react-number-field/#hook)
+ *
+ * API:
+ *
+ * - [useNumberField API](https://mui.com/base-ui/react-number-field/hooks-api/#use-number-field)
+ */
+export function useNumberField(params: NumberFieldProps): UseNumberFieldReturnValue {
+  const {
+    id: idProp,
+    name,
+    min,
+    max,
+    smallStep = 1,
+    step = 1,
+    largeStep = 10,
+    required = false,
+    disabled = false,
+    invalid = false,
+    readOnly = false,
+    autoFocus = false,
+    allowWheelScrub = false,
+    format,
+    value: externalValue,
+    onChange,
+    defaultValue,
+  } = params;
+
+  const minWithDefault = min ?? Number.MIN_SAFE_INTEGER;
+  const maxWithDefault = max ?? Number.MAX_SAFE_INTEGER;
+  const minWithZeroDefault = min ?? 0;
+  const formatStyle = format?.style;
+
+  const allowedNonNumericKeys = React.useMemo(() => {
+    const value = ['.', ','];
+    if (formatStyle === 'percent') {
+      value.push('%');
+    }
+    if (minWithDefault && minWithDefault <= 0) {
+      value.push('-');
+    }
+    // Consider adding more allowed keys, such as for currencies.
+    return value;
+  }, [minWithDefault, formatStyle]);
+
+  const id = useId(idProp);
+
+  const formatOptionsRef = useLatestRef(format);
+
+  const inputRef = React.useRef<HTMLInputElement>(null);
+  const scrubAreaRef = React.useRef<HTMLSpanElement>(null);
+
+  const scrubHandleRef = React.useRef<{
+    direction: 'vertical' | 'horizontal';
+    onScrub: (event: PointerEvent) => void;
+    onScrubbingChange: (isScrubbing: boolean, event: PointerEvent) => void;
+  } | null>(null);
+
+  const startTickTimeoutRef = React.useRef(-1);
+  const tickIntervalRef = React.useRef(-1);
+  const intentionalTouchCheckTimeoutRef = React.useRef(-1);
+  const avoidFlickerTimeoutRef = React.useRef(-1);
+  const isPressedRef = React.useRef(false);
+  const isHoldingShiftRef = React.useRef(false);
+  const isHoldingMetaRef = React.useRef(false);
+  const isScrubbingRef = React.useRef(false);
+  const incrementDownCoordsRef = React.useRef({ x: 0, y: 0 });
+  const movesAfterTouchRef = React.useRef(0);
+  const allowInputSyncRef = React.useRef(true);
+  const unsubscribeFromGlobalContextMenuRef = React.useRef<() => void>(() => {});
+  const isTouchingRef = React.useRef(false);
+
+  const [value, setValueUnwrapped] = useControlled<number | null>({
+    controlled: externalValue,
+    default: defaultValue,
+    name: 'NumberField',
+    state: 'value',
+  });
+
+  // We need to hide the input value during SSR to avoid a potential mismatch between the server
+  // rendered value and the client rendered value when their locales differ.
```

Comment:
Oh, maybe we could still render it but with `suppressHydrationWarning`? I guess we only care about this because of the warning but that warning is useless in our case?

Reply:
There'll be a flash regardless; it depends which flash is better:

- Flash from empty field to formatted field regardless if there's a match or not
- Flash from server's formatting to the user's one. If the server matches the client, there's no flash. This indeed may be better.

Reply:
> Flash from server's formatting to the user's one. If the server matches the client, there's no flash. This indeed may be better.

👍